src/ooni/web_connectivity.cpp

```diff
@@ -0,0 +1,677 @@
+      // TODO set the appropriate headers to support cloud-fronting.
+    }
+
+    logger->debug("web_connectivity: performing control request to %s",
+                  options["backend"].c_str());
+    http::request(request_settings, headers, body,
+            [=](Error error, Var<http::Response> response) {
+      if (!error) {
+        try {
+          (*entry)["control"] = Entry::parse(response->body);
+          callback(NoError());
+          return;
+        } catch (const std::invalid_argument &) {
+          (*entry)["control_failure"] = "json_parse_error";
+          callback(JsonParseError());
+          return;
+        }
+      }
+      (*entry)["control_failure"] = error.as_ooni_error();
+      callback(error);
+      return;
+    });
+}
+
+static void experiment_http_request(Var<Entry> entry,
+        std::string url, Callback<Error> cb,
+        Settings options,
+        Var<Reactor> reactor, Var<Logger> logger) {
+
+  http::Headers headers = COMMON_CLIENT_HEADERS;
+  std::string body;
+  options["http/url"] = url;
+
+  logger->debug("Requesting url %s", url.c_str());
+  templates::http_request(entry, options, headers, body, [=](Error err,
+              Var<http::Response> response) {
+    if (err) {
+      (*entry)["http_experiment_failure"] = err.as_ooni_error();
+      cb(err);
+      return;
+    }
+    cb(NoError());
+  }, reactor, logger);
+}
+
+static void experiment_tcp_connect(Var<Entry> entry,
+        SocketList sockets, Callback<Error> cb,
+        Var<Reactor> reactor, Var<Logger> logger) {
+
+  int socket_count = sockets.size();
+  Var<int> sockets_tested(new int(0));
+  // XXX this is very ghetto
+  if (socket_count == 0) {
+    cb(NoError());
+  }
+
+  auto handle_connect = [=](std::string ip, int port) {
+    return [=](Error err, Var<net::Transport> txp) {
+      *sockets_tested += 1;
+      bool close_txp = true;
+      Entry result = {
+        {"ip", ip},
+        {"port", port},
+        {"status", {
+            {"success", nullptr},
+            {"failure", nullptr},
+            {"blocked", nullptr}
+          }
+        },
+      };
+
+      if (!!err) {
+          logger->info("web_connectivity: failed to connect to %s:%d",
+                       ip.c_str(), port);
+          result["status"]["success"] = false;
+          result["status"]["failure"] = err.as_ooni_error();
+          close_txp = false;
+      } else {
+          logger->info("web_connectivity: success to connect to %s:%d",
+                       ip.c_str(), port);
+          result["status"]["success"] = true;
+          result["status"]["blocked"] = false;
+      }
+      (*entry)["tcp_connect"].push_back(result);
+      if (socket_count == *sockets_tested) {
+        if (close_txp == true) {
+          txp->close([=]{
+            cb(NoError());
+          });
+        } else {
+          cb(NoError());
+        }
+      } else {
+        if (close_txp == true) {
+          // XXX optimistic closure
```

Comment:
What I mean is that I don't actually wait for the connection to be closed, but call close on it and expect everything to go well.

I guess since closing a socket doesn't require bidirectional communication there shouldn't be much that can go wrong, though I wanted to take note of the fact that I am not handling any errors or waiting for the callback to actually fire.